frontend/ui/screensaver.lua

```diff
@@ -60,93 +37,204 @@ local function getRandomImage(dir)
-            height = Screen:getHeight(),
-            width = Screen:getWidth(),
-            scale_factor = doc_settings:readSetting("proportional_screensaver") and 0 or nil,
-        }
-        return createWidgetFromImage(img_widget)
+    local exclude_ss = doc_settings:readSetting("exclude_screensaver")
+    doc_settings:close()
+    return exclude_ss or false
+end
+
+function Screensaver:setMessage()
+    local InputDialog = require("ui/widget/inputdialog")
+    local screensaver_message = G_reader_settings:readSetting("screensaver_message")
+    if screensaver_message == nil then
+        screensaver_message = defalut_screensaver_message
     end
+    self.input_dialog = InputDialog:new{
+        title = "Screensaver message",
+        input = screensaver_message,
+        buttons = {
+            {
+                {
+                    text = _("Cancel"),
+                    callback = function()
+                        UIManager:close(self.input_dialog)
+                    end,
+                },
+                {
+                    text = _("Set message"),
+                    is_enter_default = true,
+                    callback = function()
+                        G_reader_settings:saveSetting("screensaver_message", self.input_dialog:getInputText())
+                        UIManager:close(self.input_dialog)
+                    end,
+                },
+            },
+        },
+    }
+    self.input_dialog:onShowKeyboard()
+    UIManager:show(self.input_dialog)
 end
 
-function Screensaver:show(kind, default_msg)
-    logger.dbg("show screensaver")
-    local InfoMessage = require("ui/widget/infomessage")
-    local screensaver_settings = G_reader_settings:readSetting(kind .. "_screensaver") or {}
-    -- first check book cover image, on by default
-    local screensaver_last_book = screensaver_settings.use_last_file or
-        G_reader_settings:readSetting("use_lastfile_as_screensaver")
-    if screensaver_last_book == nil or screensaver_last_book then
+function Screensaver:show()
+    local screensaver_type = G_reader_settings:readSetting("screensaver_type")
+    local widget = nil
+    self.left_msg = nil
+    if screensaver_type == "cover" then
         local lastfile = G_reader_settings:readSetting("lastfile")
-        if lastfile then
-            local doc_settings = DocSettings:open(lastfile)
-            local exclude = doc_settings:readSetting("exclude_screensaver")
-            if not exclude then
-                self.left_msg = self:getCoverImage(lastfile)
-            end
+        local doc_settings = DocSettings:open(lastfile)
+        local exclude = doc_settings:readSetting("exclude_screensaver")
+        if exclude ~= true then
+            local doc = DocumentRegistry:openDocument(lastfile)
+            local image = doc:getCoverPageImage()
+            doc:close()
+            widget = ImageWidget:new{
+                image = image,
+                image_disposable = false,
```

Comment:
Should the `image` you just got really be _not_ disposable (default is `image_disposable=true`, which means ImageWidget will do the image memory cleanup). If it should, I don't see you calling `:free()` on it to clean it up anywhere.